### PR TITLE
feat: Media Library support (Android)

### PR DESCRIPTION
**Pull request by dcalhoun**
## Related

* https://github.com/Automattic/dotcom-forge/issues/9086
* https://github.com/wordpress-mobile/WordPress-Android/pull/21399

## Description 

Add support for attaching Media Library items to media blocks—Image, Gallery, Video, etc. 

## Testing Instructions

> [!TIP]
> Use the WPAndroid [prototype build](https://github.com/wordpress-mobile/WordPress-Android/pull/21399#issuecomment-2452715659) for testing. 

**1 - Inserting media**
1. Add a media-related blocks—Image, Gallery, etc. 
2. Attach media from the Media Library. 
3. Verify the media displays as expected. 

**2 - Modifying media**
1. Open a post with an existing Gallery block. 
2. Select an image within the gallery. 
3. Tap _Replace_ and select a different image from the Media Library. 
4. Verify the image is replaced as expected. 

**3 - Add media**
1. Open a post with an existing Gallery block. 
2. Select an image within the gallery. 
3. Select the parent Gallery block by tapping the three vertical dots at the end of the keyboard toolbar. 
4. Tap _Add_. 
5. Add additional images from the Media Library. 
6. Verify the images are added as expected. 

**4 - Cancel additions/modifications**
1. Open a post with an existing Gallery block. 
2. Select an image within the gallery. 
3. Select the parent Gallery block by tapping the three vertical dots at the end of the keyboard toolbar. 
4. Tap _Add_. 
5. Tap _Open Media Library_. 
6. Unselect images, select new images. 
7. Tap the back arrow icon. 
8. Tap the X icon.
9. Verify gallery remains unchanged.

### CODE REVIEW

File: Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt [hunk 1]
```diff
@@ -351,3 +426,74 @@ object GutenbergWebViewPool {
         preloadedWebView = null
     }
 }
+
+data class Media(
```
Comment (dcalhoun) [inline Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt:430]: The `Media` class was copied from the [Gutenberg Mobile bridge](https://github.com/WordPress/gutenberg/blob/ee2fb2925df35fc141e0ecc86b907f0cf6a18cef/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/WPAndroidGlue/Media.kt#L12-L70) and adapted.

File: Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt [hunk 2]
```diff
@@ -351,3 +426,74 @@ object GutenbergWebViewPool {
...
+            alt: String?,
+            metadata: Bundle = Bundle()
+        ): Media {
+            val type = convertToType(mimeType)
+            return Media(id, url, type, caption ?: "", title ?: "", alt ?: "", metadata)
+        }
+    }
+}
+
+enum class MediaType(var label: String) {
```
Comment (dcalhoun) [inline Demo-Android/Gutenberg/src/main/java/org/wordpress/gutenberg/GutenbergView.kt:480]: The `MediaType` enum was copied from the [Gutenberg Mobile bridge](https://github.com/WordPress/gutenberg/blob/ee2fb2925df35fc141e0ecc86b907f0cf6a18cef/packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergBridgeJS2Parent.java#L94-L117) and adapted.

File: ReactApp/src/misc/Helpers.js [hunk 3]
```diff
@@ -59,7 +59,7 @@ export function blurEditor() {
 
 export function openMediaLibrary(config) {
 	if (window.editorDelegate) {
-		window.editorDelegate.openMediaLibrary(config);
+		window.editorDelegate.openMediaLibrary(JSON.stringify(config));
```
Comment (dcalhoun) [inline ReactApp/src/misc/Helpers.js:62]: In contrast to iOS—maybe I overlooked something—I could not identify a way to send a literal object from Android to the WebView. I was required to stringify and decode a JSON object  instead.

**Comment by nbradbury** [comment]
@dcalhoun Nice job, this is all working as expected! I do have a question about the last testing step: you mentioned tapping _Cancel_ but I don't see a cancel button anywhere. Where should the cancel button be?

Also, I'm seeing a crash which I believe is unrelated to this PR but wanted to mention it:

* Select a photo in a gallery
* Tap Replace
* Tap the URL under "Current media Url"
* Tap the back button
* Boom!

```
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:558)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
                                                                                                    Caused by: java.lang.reflect.InvocationTargetException
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936) 
                                                                                                    Caused by: org.json.JSONException: Value null of type org.json.JSONObject$1 cannot be converted to JSONObject
                                                                                                    	at org.json.JSON.typeMismatch(JSON.java:112)
                                                                                                    	at org.json.JSONObject.<init>(JSONObject.java:172)
                                                                                                    	at org.json.JSONObject.<init>(JSONObject.java:185)
                                                                                                    	at org.wordpress.gutenberg.GutenbergView.getTitleAndContent$lambda$1$lambda$0(GutenbergView.kt:288)
```

**Comment by dcalhoun** [comment]
> I do have a question about the last testing step: you mentioned tapping _Cancel_ but I don't see a cancel button anywhere. Where should the cancel button be?

@nbradbury I left out a couple of steps before arriving to the ability to cancel modifications, I meant pressing the "X" button for the dialog after opening the _Media Library_. I updated the test case 4's description. 

> Also, I'm seeing a crash which I believe is unrelated to this PR but wanted to mention it:

Thank you for noting this. It does appear unrelated to these changes. I created https://github.com/Automattic/dotcom-forge/issues/9828 to address it separately. 

One last request: if you have time, you could "smoke test" the Gutenberg Mobile (disable experimental editor feature flags) and Aztec (P9ugOq-4rN-p2) editors to verify no regressions arose. I did so myself, but welcome additional testing.

**Comment by nbradbury** [comment]
> One last request: if you have time, you could "smoke test" the Gutenberg Mobile (disable experimental editor feature flags) and Aztec (P9ugOq-4rN-p2) editors to verify no regressions arose.

Funny, until I read this I didn't know we still supported Aztec :) I'll test both Gutenberg Mobile and Aztec and report my findings.

**Comment by nbradbury** [comment]
@dcalhoun Tested with both Gutenberg Mobile and Aztec and didn't run into any surprises. Do the [two code scanning errors](https://github.com/wordpress-mobile/GutenbergKit/pull/36/checks?check_run_id=32893885966) need to be addressed?

**Comment by dcalhoun** [comment]
> Do the [two code scanning errors](https://github.com/wordpress-mobile/GutenbergKit/pull/36/checks?check_run_id=32893885966) need to be addressed?

@nbradbury no, I do not believe they need to be addressed directly. It's a case of a scanner flagging a minified build. 

The [errors appear](https://github.com/wordpress-mobile/GutenbergKit/security/code-scanning/38) to relate to string replacement. The only `replace` reference _not_ utilizing regular expression appears to [originate from third-party code](https://github.com/alexei/sprintf.js/blob/3a0d8c26d291b5bd9f1974877ecc50739921d6f5/src/sprintf.js#L127)—the ubiquitous `sprintf`—which is utilized by [`@wordpress/i18n`](https://github.com/WordPress/gutenberg/blob/e7f0dc53ab3d8bbd53bd19129e00ed34aad67326/packages/i18n/package.json#L38) that is [bundled with build output](https://github.com/wordpress-mobile/GutenbergKit/blob/6cc307e7fc24910697be5f71b7d70f465a9c0f63/ReactApp/vite.config.remote.js#L50-L55). We plan to remove this minified build file from version control in the future; its presence is solely for CI build task purposes.

**Review by nbradbury — APPROVED** [review]
:shipit:

**Pull request merged**